app/services/bulk_hosts_manager.rb

```diff
@@ -38,4 +38,19 @@ def rebuild_configuration
     end
     all_fails
   end
+
+  def assign_taxonomy(taxonomy, optimistic_import)
+    type = taxonomy.type.downcase
+    if optimistic_import == 'yes'
+      @hosts.update_all("#{type}_id".to_sym => taxonomy.id)
+      # hosts location needs to be updated before import missing ids
+      # taxonomy.import_missing_ids
+    else
+      if taxonomy.need_to_be_selected_ids.count == 0
+        # @hosts.update_all("#{type}_id".to_sym => taxonomy.id)
```

Comment:
Seems this piece of code has been broken for a while.  It is refactored here to keep the record.